src/framework/commandManager/commandRegistry.ts

```diff
@@ -0,0 +1,213 @@
+    await this.loadCommandsRecursively(folderPath);
+  }
+
+  private async loadCommandsRecursively(folderPath: string): Promise<void> {
+    const items = readdirSync(folderPath, { withFileTypes: true });
+
+    await Promise.all(
+      items.map(async (item) => {
+        const fullPath = path.join(folderPath, item.name);
+
+        if (item.isDirectory()) {
+          await this.loadCommandsRecursively(fullPath);
+        } else if (
+          item.isFile() &&
+          CommandRegistry.ValidExtensions.includes(path.extname(item.name))
+        ) {
+          const commands = await this.importCommand(fullPath);
+          commands.forEach((command) => this.register(command));
+        }
+      }),
+    );
+  }
+
+  private async importCommand(filePath: string): Promise<BaseCommand[]> {
+    this.bot.logger.debug(`Importing commands from ${filePath}`);
+    try {
+      // Convert to file URL for proper module resolution
+      const importPath = `file://${filePath}`;
+      const module = await import(importPath);
+
+      if (!this.isValidModule(module)) {
+        this.bot.logger.warn(
+          `Skipping invalid module at ${filePath}. Expected a valid ES module, got ${typeof module}`,
+        );
+        return [];
+      }
+
+      const commands: BaseCommand[] = [];
+
+      const potentialCommands = Object.values(module);
+      for (const exportedValue of potentialCommands) {
+        if (!isLichobiCommandConstructor(exportedValue)) {
+          continue;
+        }
+        try {
+          const command = new exportedValue(this.bot);
+          commands.push(command);
+        } catch (error) {
+          this.bot.logger.error(
+            `Failed to instantiate command from ${filePath}:`,
+            error,
+          );
+        }
+      }
+
+      if (commands.length === 0) {
+        this.bot.logger.warn(`No valid commands found in ${filePath}`);
+        return [];
+      }
+
+      return commands;
+    } catch (error) {
+      this.bot.logger.error(`Error loading commands from ${filePath}:`, error);
+      return [];
+    }
+  }
+
+  private isValidModule(module: unknown): module is Record<string, unknown> {
+    return (
+      typeof module === "object" &&
+      module !== null &&
+      !Array.isArray(module) &&
+      Object.entries(module as object).length > 0
+    );
+  }
+
+  public async registerApplicationCommandsOnDiscord(
+    guildId?: Snowflake,
+  ): Promise<void> {
+    const commands: ApplicationCommandDataResolvable[] = [];
+    for (const command of this.chatInputCommands.values()) {
+      commands.push({
+        ...command.getBaseCommandData(),
+        ...command.getChatInputCommandData(),
+      });
+    }
+    for (const command of this.messageContextMenuCommands.values()) {
+      commands.push(command.getBaseCommandData());
+    }
+    for (const command of this.userContextMenuCommands.values()) {
+      commands.push(command.getBaseCommandData());
+    }
```

Comment:
let's make this a method in mixin class to get this data

Reply:
mixin doesn't have access to base, so this is not possible. Doesn't make sense to move to base class. Let registry handle this.